src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/MatchResult.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.PatternMatching;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion
+{
+    internal struct MatchResult
+    {
+        public readonly RoslynCompletionItem RoslynCompletionItem;
+        public readonly bool MatchedFilterText;
+        public readonly ImmutableArray<Span> HighlightedSpans;
+
+        // In certain cases, there'd be no match but we'd still set `MatchedFilterText` to true,
+        // e.g. when the item is in MRU list. Therefore making this nullable.
+        public readonly PatternMatch? PatternMatch;
+
+        public readonly VSCompletionItem VSCompletionItem;
+
+        // We want to preserve the original alphabetical order for items with same pattern match score,
+        // but `ArrayBuilder.Sort` we currently use isn't stable. So we have to add a monotonically increasing 
+        // integer to archieve this.
+        private readonly int _indexInOriginalSortedOrder;
+
+        public MatchResult(
+            RoslynCompletionItem roslynCompletionItem, VSCompletionItem vsCompletionItem,
+            bool matchedFilterText, PatternMatch? patternMatch, int index,
+            ImmutableArray<Span> highlightedSpans)
+        {
+            RoslynCompletionItem = roslynCompletionItem;
+            MatchedFilterText = matchedFilterText;
+            PatternMatch = patternMatch;
+            VSCompletionItem = vsCompletionItem;
+            _indexInOriginalSortedOrder = index;
+            HighlightedSpans = highlightedSpans;
+        }
+
+        public static IComparer<MatchResult> Comparer => ExtendedFilterResultComparer.Instance;
+
+        private class ExtendedFilterResultComparer : IComparer<MatchResult>
```

Comment:
It might be confusing we have multiple way to compare MatchResult, consider adding comments to explain (I think one is for sorting, another is for selection, right?)